Source/WebCore/crypto/gcrypt/CryptoKeyOKPGCrypt.cpp

```diff
@@ -63,52 +63,77 @@ static bool supportedAlgorithmIdentifier(CryptoAlgorithmIdentifier keyIdentifier
     return false;
 }
 
+Vector<uint8_t> nine {
+    0x09, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+};
+
 }
 
-std::optional<CryptoKeyPair> CryptoKeyOKP::platformGeneratePair(CryptoAlgorithmIdentifier identifier, NamedCurve namedCurve, bool extractable, CryptoKeyUsageBitmap usages)
+static void gcryptGenerateEd25519Keys(Vector<uint8_t>&& publicKeyData, Vector<uint8_t>&& privateKeyData)
 {
-    if (namedCurve != NamedCurve::Ed25519 && namedCurve != NamedCurve::X25519)
-        return { };
-
     PAL::GCrypt::Handle<gcry_sexp_t> genkeySexp;
-    gcry_error_t error = gcry_sexp_build(&genkeySexp, nullptr, namedCurve == CryptoKeyOKP::NamedCurve::Ed25519 ? "(genkey (ecdsa (curve Ed25519) (flags eddsa)))" : "(genkey (ecc (curve Curve25519)))");
+    gcry_error_t error = gcry_sexp_build(&genkeySexp, nullptr, "(genkey (ecdsa (curve Ed25519) (flags eddsa)))");
     if (error != GPG_ERR_NO_ERROR) {
         PAL::GCrypt::logError(error);
-        return std::nullopt;
+        return;
     }
 
     PAL::GCrypt::Handle<gcry_sexp_t> keyPairSexp;
     error = gcry_pk_genkey(&keyPairSexp, genkeySexp);
     if (error != GPG_ERR_NO_ERROR) {
         PAL::GCrypt::logError(error);
-        return std::nullopt;
+        return;
     }
 
     PAL::GCrypt::Handle<gcry_mpi_t> qMpi, dMpi;
     error = gcry_sexp_extract_param(keyPairSexp, "private-key", "qd", &qMpi, &dMpi, nullptr);
     if (error != GPG_ERR_NO_ERROR) {
         PAL::GCrypt::logError(error);
-        return std::nullopt;
+        return;
     }
 
-    // Retrieve the `q` data, which should be in the uncompressed point format for the X25519 keys.
-    // Validate the data size and the first byte (which should be 0x40).
-    size_t pubKeySize = namedCurve == NamedCurve::Ed25519 ? 32 : 33;
-    auto publicKeyData = mpiData(qMpi);
-    if (!publicKeyData || publicKeyData->size() != pubKeySize
-        || (namedCurve == NamedCurve::X25519
-            && !CryptoConstants::matches(publicKeyData->data(), 1, CryptoConstants::s_x25519UncompressedFormatLeadingByte))) {
+    publicKeyData = *mpiData(qMpi);
+    privateKeyData = *mpiData(dMpi);
+}
+
+static void gcryptGenerateX25519Keys(Vector<uint8_t>&& publicKeyData, Vector<uint8_t>&& privateKeyData)
+{
+    // private key is just 32 random bytes
+    PAL::GCrypt::Handle<gcry_mpi_t> mpi(gcry_mpi_new(256));
+    gcry_mpi_randomize(mpi, 256, GCRY_STRONG_RANDOM);
+    auto q = mpiData(mpi);
+
+    // public key being X25519(a, 9), as defined in [RFC7748], section 6.1.
+    auto d = X25519(*q, WebCore::CryptoKeyOKPImpl::nine);
+
+    publicKeyData = *q;
+    privateKeyData = *d;
+}
+
+std::optional<CryptoKeyPair> CryptoKeyOKP::platformGeneratePair(CryptoAlgorithmIdentifier identifier, NamedCurve namedCurve, bool extractable, CryptoKeyUsageBitmap usages)
+{
+    Vector<uint8_t> publicKeyData, privateKeyData;
```

Comment:
This is then `std::optional<std::pair<Vector<uint8_t>, Vector<uint8_t>>> keyPair;'. By default it's nullopt.